### DISCUSSION TITLE
Recording ephemeral output in Artifact Input Manifest

### DISCUSSION THREAD

**Discussion by bharsesh** [omnibor/spec#47] [General]
During the build process, output could be produced that is ephemeral. Should such output be captured in the OmniBOR document a.k.a Artifact Input Manifest (AIM)  ?
For example,  consider the build command 'gcc -c foo.c -o foo.o'
The above command implicitly does 
 cc foo.c -o tmp-foo.s
 as tmp-foo.s -o foo.o 
If we enable OMNIBOR_DIR environment variable, the compiler front-end and assembler becomes OmniBOR aware and we may potentially end up creating AIM for the output corresponding to tmp-foo.s and foo.o unless the assembler is taught NOT to create AIM or embed OmniBOR in certain situations. (Note that If OmniBOR was enabled by command line options instead of an environment variable, then we could avoid creating AIM for tmp-foo.s).  Once the gcc command completes execution, the output file tmp-foo.s is not guaranteed to exist.  So should AIM include that as a dependency for the output file foo.o ? 
Note that tmp-foo.s will correspond to a non-leaf artifact in the AIM, the leaf artifact (foo.c) would still be persistent.  

In general, the questions/points discussed so far are:
- Should AIM contain ephemeral artifacts? 
-  Should we say ephemeral non-leaf artifacts are allowed in the AIM but all leaf artifacts should be non-ephemeral?
- Can the OmniBOR spec provide guidance on this topic as opposed to a requirement or would this be an implementation dependent aspect of OmniBOR supporting tools?
- Should transient files be tagged appropriately so consumer tools either strip or ignore them?
- Should every OmniBOR document either be able to be presented or "witnessed"?

**Comment by bharsesh** [comment]
Here's a simple proposal that can provide the choice and flexibility for build tool implementers to capture temporary output. 

The spec could provide a general guidance like the following -
"Every" entry in the OmniBOR document should adhere to this rule:

- A _blob_ entry in the OmniBOR document should be non-ephemeral (OR) If a _blob_ entry is ephemeral, then it MUST have a link to its OmniBOR document following the _bom_ keyword and this document MUST be non-ephemeral.
Example:
_blob_ file // file MUST be non-ephemeral.
_blob_  file _bom_ OmniBOR_ID_file // file may be ephemeral but OmniBOR_ID_file MUST be non-ephemeral.
-----------------

**Reply by alilleybrinker** [reply]
1. We probably want a term for "non-ephemeral" artifacts that's not a negation. Maybe "realized"? Happy to workshop this.
2. Do we want a way to explicitly _say_ that something is ephemeral, or do we only consider things to be permissibly implicitly ephemeral if they're non-leaf (meaning: if they have an AIM via a `bom` reference).

**Reply by edwarnicke** [reply]
Attempt at summary of discussion:

Any given build tool being enlightened to support OmniBOR data cannot know whether the output it produces is ephemeral.

Any given build tool should minimally encode any output it has persisted beyond the conclusion of its invocation.
